lib/class-wp-rest-block-directory-controller.php

```diff
@@ -0,0 +1,311 @@
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function install_block( $request ) {
+
+		include_once( ABSPATH . 'wp-admin/includes/file.php' );
+		include_once( ABSPATH . 'wp-admin/includes/plugin.php' );
+		include_once( ABSPATH . 'wp-admin/includes/class-wp-upgrader.php' );
+		include_once( ABSPATH . 'wp-admin/includes/plugin-install.php' );
+
+		$api = plugins_api(
+			'plugin_information',
+			array(
+				'slug'   => $request->get_param( 'slug' ),
+				'fields' => array(
+					'sections' => false,
+				),
+			)
+		);
+
+		if ( is_wp_error( $api ) ) {
+			return WP_Error( $api->get_error_code(), $api->get_error_message() );
+		}
+
+		$skin     = new WP_Ajax_Upgrader_Skin();
+		$upgrader = new Plugin_Upgrader( $skin );
+
+		$filesystem_method = get_filesystem_method();
+
+		if ( 'direct' !== $filesystem_method ) {
+			return WP_Error( null, 'Only direct FS_METHOD is supported.' );
+		}
+
+		$result = $upgrader->install( $api->download_link );
+
+		if ( is_wp_error( $result ) ) {
+			return WP_Error( $result->get_error_code(), $result->get_error_message() );
+		}
+
+		if ( is_wp_error( $skin->result ) ) {
+			return WP_Error( $skin->$result->get_error_code(), $skin->$result->get_error_message() );
+		}
+
+		if ( $skin->get_errors()->has_errors() ) {
+			return WP_Error( $skin->$result->get_error_code(), $skin->$result->get_error_messages() );
+		}
+
+		if ( is_null( $result ) ) {
+			global $wp_filesystem;
+			// Pass through the error from WP_Filesystem if one was raised.
+			if ( $wp_filesystem instanceof WP_Filesystem_Base && is_wp_error( $wp_filesystem->errors ) && $wp_filesystem->errors->has_errors() ) {
+				return WP_Error( 'unable_to_connect_to_filesystem', esc_html( $wp_filesystem->errors->get_error_message() ) );
+			}
+			return WP_Error( 'unable_to_connect_to_filesystem', __( 'Unable to connect to the filesystem. Please confirm your credentials.', 'gutenberg' ) );
+		}
+
+		$install_status = install_plugin_install_status( $api );
+
+		$activate_result = activate_plugin( $install_status['file'] );
+
+		if ( is_wp_error( $activate_result ) ) {
+			return WP_Error( $activate_result->get_error_code(), $activate_result->get_error_message() );
+		}
+
+		return rest_ensure_response( true );
+	}
+
+	/**
+	 * Deactivates and deletes a plugin
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function uninstall_block( $request ) {
+
+		include_once( ABSPATH . 'wp-admin/includes/file.php' );
+		include_once( ABSPATH . 'wp-admin/includes/plugin.php' );
+		include_once( ABSPATH . 'wp-admin/includes/class-wp-upgrader.php' );
+		include_once( ABSPATH . 'wp-admin/includes/plugin-install.php' );
+
+		$api = plugins_api(
```

Comment:
We can _probably_ simplify this function to remove the need to query the API, providing some assumptions are true:

If the `uninstall_block()` method is only ever used to uninstall blocks that were installed using the `install_block()` method, *and* the slug is the plugin slug (which means it's also the folder name), then we can skip more or less straight to the `deactivate_plugins()` step.